Cpp/Library/Utility/Exception.cc

```diff
@@ -22,53 +22,34 @@ namespace Nymph
         {
             std::rethrow_if_nested( e );
         }
-        catch(const Nymph::Exception& e)
+        catch(const scarab::base_exception& e)
+        {
+            PrintException( e, ++count );
+        }
+        catch(const std::exception& e)
         {
             PrintException( e, ++count );
         }
         return;
     }
 
-/*
-    BaseException::BaseException() :
-            std::exception(),
-            fAtFilename( "unknown" ),
-            fAtLineNumber( 0 ),
-            fBuffer()
-    {}
-
-    BaseException::~BaseException() noexcept
-    {}
-
-    const char* BaseException::where() const noexcept
+    void PrintException( const std::exception& e, unsigned count )
```

Comment:
I'm a little confused why the `PrintException` function here is overloaded for both types of exception types while also giving catches for both types of exception classes. Is this redundant/unnecessary, or are there instances where an exception is nested across these different types?

Reply:
That's a good question.  The reason has to do with why the recursive print function is setup as it is, which has to do with the use of nested exceptions.  As I note way down below in this explanation, I do see something that's wrong that needs fixing to make the code clearer.

In each `PrintException()` function, at the beginning of the function and up until `std::rethrow_if_nested()` is called, the variable `e` is of the type relevant to the particular overload of `PrintException()`.  When the exception is rethrown, it's rethrown as the next type down in the nesting.  So we need to catch both types so we can call the appropriate overload of `PrintException()` for that next nesting level.  If we only caught as `std::exception()`, the function overload resolution would not detect that some exceptions inherit from `scarab::base_exception`, while others do not.

The order of the `catch` statements is important: we need to first attempt to catch as `scarab::base_exception` since that class inherits from `std::exception`.  If we did it in the opposite order, it'd always get caught as `std::exception`, and we'd never see the `where()` information.

The problem I now notice in the code is that I use the same variable name to catch the exception as was used in the `PrintException()` argument, `e`.  While it does work, it makes the code unclear.  I'll catch with a different variable name.

Reply:
Ah that makes sense. Thanks for the response and for updating the variable name for clarity!